src/Processors/QueryPlan/ReadFromRemote.cpp

```diff
@@ -312,28 +317,33 @@ void ReadFromParallelRemoteReplicasStep::enforceAggregationInOrder()
 void ReadFromParallelRemoteReplicasStep::initializePipeline(QueryPipelineBuilder & pipeline, const BuildQueryPipelineSettings &)
 {
     Pipes pipes;
-
     const Settings & current_settings = context->getSettingsRef();
     auto timeouts = ConnectionTimeouts::getTCPTimeoutsWithFailover(current_settings);
 
-    for (size_t replica_num = 0; replica_num < shard.shard_info.getAllNodeCount(); ++replica_num)
+    size_t replica_num = 0;
+    while (pipes.size() != current_settings.max_parallel_replicas - 1)
```

Comment:
This is needed, because the value of the setting `max_parallel_replicas` could be less than the total number of replicas in a shard. And we have to pick only remote replicas with remote addresses. This loop is a little bit difficult to read, yes...